plugin/path_reserves.go

```diff
@@ -0,0 +1,312 @@
+		},
+	}, nil
+}
+
+func (b *backend) operationReserveDelete(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	reserveName := fieldData.Get("name").(string)
+	reserve, err := readReserve(ctx, req.Storage, reserveName)
+	if err != nil {
+		return nil, err
+	}
+	if reserve == nil {
+		return nil, nil
+	}
+	// We need to remove all the passwords we'd stored for these service accounts.
+	var deletionErrs error
+	for _, serviceAccountName := range reserve.ServiceAccountNames {
+		if err := b.deleteReserveServiceAccount(ctx, req.Storage, serviceAccountName); err != nil {
+			deletionErrs = multierror.Append(deletionErrs, err)
+		}
+	}
+	if deletionErrs != nil {
+		return nil, deletionErrs
+	}
+	if err := req.Storage.Delete(ctx, libraryPrefix+reserveName); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+// readReserve is a helper method for reading a reserve from storage by name.
+// It's intended to be used anywhere in the plugin. It may return nil, nil if
+// a libraryReserve doesn't currently exist for a given reserveName.
+func readReserve(ctx context.Context, storage logical.Storage, reserveName string) (*libraryReserve, error) {
+	entry, err := storage.Get(ctx, libraryPrefix+reserveName)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+	reserve := &libraryReserve{}
+	if err := entry.DecodeJSON(reserve); err != nil {
+		return nil, err
+	}
+	return reserve, nil
+}
+
+// storeReserve is only intended to be used inside path_reserves.go, because this is the only place
+// that is intended to be editing reserves.
+func storeReserve(ctx context.Context, storage logical.Storage, reserveName string, reserve *libraryReserve) error {
+	entry, err := logical.StorageEntryJSON(libraryPrefix+reserveName, reserve)
+	if err != nil {
+		return err
+	}
+	return storage.Put(ctx, entry)
+}
+
+// ensureNotInAnotherReserve is a helper method for checking that new service accounts aren't already being
+// managed by another reserve. This is because we would experience a lot of unexpected behavior if the same
+// service account were being checked in (with its password rolled) at different times through another
+// reserve, and it could be difficult to debug because it'd be happening outside the times we'd be looking
+// at in the logs.
+func ensureNotInAnotherReserve(ctx context.Context, storage logical.Storage, newServiceAccountNames []string) error {
+	// Gather up all the service accounts currently being managed.
+	preExistingReserveNames, err := storage.List(ctx, libraryPrefix)
+	if err != nil {
+		return err
+	}
+	preExistingServiceAccountNames := make(map[string]bool)
+	for _, preExistingReserveName := range preExistingReserveNames {
+		preExistingReserve, err := readReserve(ctx, storage, preExistingReserveName)
+		if err != nil {
+			return err
+		}
+		for _, preExistingServiceAccountName := range preExistingReserve.ServiceAccountNames {
```

Comment:
I ended up deleting that function altogether. FWIW, though, in Go if you try to range over a nil array, it will be fine. https://play.golang.org/p/anrTp2quLHh